src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,254 @@
+#include "firmware/app/control/trajectory_planner.h"
+
+#include "firmware/shared/math/polynomial_1d.h"
+#include "firmware/shared/math/polynomial_2d.h"
+#include "firmware/shared/math/vector_2d.h"
+#include "math.h"
+#include "stdio.h"
+
+void generate_constant_arc_length_segmentation(
+    FirmwareRobotPathParameters_t path_parameters, Trajectory_t* trajectory)
+{
+    TrajectoryElement_t reverse_trajectory[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+
+    double max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    double velocity_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+
+
+    // Set the internal path parameter variables
+    TrajectoryElement_t* traj_elements = trajectory->trajectory_elements;
+
+    double final_speed                      = path_parameters.final_speed;
+    double initial_speed                    = path_parameters.initial_speed;
+    const double max_allowable_acceleration = path_parameters.max_allowable_acceleration;
+    const double max_allowable_speed        = path_parameters.max_allowable_speed;
+    const unsigned int num_segments         = path_parameters.num_segments;
+    const Polynomial2dOrder3_t path         = path_parameters.path;
+    double t_end                            = path_parameters.t_end;
+    double t_start                          = path_parameters.t_start;
+
+    trajectory->num_elements = path_parameters.num_segments;
+
+    // Variable used to flag if the path is moving "backwards" along the input path
+    unsigned int reverse_parameterization = 0;
+
+
+    // Check for the parameterization direction
+    // If the path is traversed in reverse, then flip all components to forwards (to be
+    // reversed again in the end)
+    if (t_end < t_start)
+    {
+        reverse_parameterization = 1;
+
+        // Reverse the direction (Polynomial library can only handle forwards direction)
+        double temp = t_start;
+        t_start     = t_end;
+        t_end       = temp;
+
+        // Reverse the final and initial speeds
+        temp          = initial_speed;
+        initial_speed = final_speed;
+        final_speed   = temp;
+    }
+
+    // Create the parmeterization to contain the desired number of segments
+    CREATE_STATIC_ARC_LENGTH_PARAMETRIZATION(arc_length_param,
+                                             TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS);
+
+    // Get all of the points for the arc length parameterization (Not constant arc length
+    // segments)
+    shared_polynomial_getArcLengthParametrizationOrder3(path, t_start, t_end,
+                                                        arc_length_param);
+
+    const double arc_segment_length =
+        arc_length_param.arc_length_values[arc_length_param.num_values - 1] /
```

Comment:
Could we add a function `double getMaxArcLength(ArcLengthParametrization_t arc_length_parametrization)`? I feel like this is more generally useful, and more importantly it's a lot easier to read.

Reply:
Made a function in trajectory_planner.c/h

Reply:
Looks like it's not here, did you push?